File: leetcode/0500_keyboard_row.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):

    ROWS = [
        set(['Q', 'W', 'E', 'R', 'T', 'Y', 'I', 'U', 'O', 'P']),
        set(['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L']),
        set(['Z', 'X', 'C', 'V', 'B', 'N', 'M']),
    ]

    def findWords(self, words):
        return [
            word for word in words if any(
                all(char in row for char in word.upper()) for row in self.ROWS)]


if __name__ == '__main__':
    solution = Solution()

    assert ['Alaska', 'Dad'] == solution.findWords(['Hello', 'Alaska', 'Dad', 'Peace'])

File: leetcode/0500_keyboard_row.kt
package leetcode

class Solution {
    val ROWS = arrayOf(
        setOf('Q', 'W', 'E', 'R', 'T', 'Y', 'I', 'U', 'O', 'P'),
        setOf('A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'),
        setOf('Z', 'X', 'C', 'V', 'B', 'N', 'M')
    )

    fun findWords(words: Array<String>): Array<String> {
        val result = mutableListOf<String>()

        for (word in words) {
            for (row in ROWS) {
                if (word.toUpperCase().all({ char -> row.contains(char) })) {
                    result.add(word)
                }
            }
        }

        return Array<String>(result.size, { i -> result[i] })
    }
}

fun main(args: Array<String>) {
    val solution = Solution()

    val words = arrayOf("Hello", "Alaska", "Dad", "Peace")

    val expected = arrayOf("Alaska", "Dad")
    val result = solution.findWords(words)

    assert(expected contentEquals result)
}
